contentctl/objects/content_versioning_service.py

```diff
@@ -0,0 +1,530 @@
+            raise KeyError(
+                "Cannot retrieve versioning status, unable to versioning status using the expected "
+                f"keys: {e}"
+            ) from e
+        raise ValueError(
+            "Cannot retrieve versioning status, unable to find an entry matching 'general' in the "
+            "response."
+        )
+
+    def activate_versioning(self) -> None:
+        """
+        Activate the content versioning service
+        """
+        # Post to the SA-ContentVersioning service to set versioning status
+        self._query_content_versioning_service(
+            method="POST",
+            body={
+                "versioning_activated": True
+            }
+        )
+
+        # Confirm versioning has been enabled
+        if not self.is_versioning_activated:
+            raise Exception("Something went wrong, content versioning is still disabled.")
+
+        self.logger.info(
+            f"[{self.infrastructure.instance_name}] Versioning service successfully activated"
+        )
+
+    @computed_field
+    @cached_property
+    def cms_fields(self) -> list[str]:
+        """
+        Property listing the fields we want to pull from the cms_main index
+
+        :returns: a list of strings, the fields we want
+        :rtype: list[str]
+        """
+        return [
+            "app_name",
+            "detection_id",
+            "version",
+            "action.correlationsearch.label",
+            "sourcetype"
+        ]
+
+    @property
+    def is_cms_parser_enabled(self) -> bool:
+        """
+        Indicates whether the cms_parser mod input is enabled or not.
+
+        :returns: a bool indicating if cms_parser mod input is activated or not
+        :rtype: bool
+        """
+        # Get the data input entity
+        cms_parser = self.service.input("data/inputs/cms_parser/main")                              # type: ignore
+
+        # Convert the 'disabled' field to an int, then a bool, and then invert to be 'enabled'
+        return not bool(int(cms_parser.content["disabled"]))                                        # type: ignore
+
+    def force_cms_parser(self) -> None:
+        """
+        Force the cms_parser to being it's run being disabling and re-enabling it.
+        """
+        # Get the data input entity
+        cms_parser = self.service.input("data/inputs/cms_parser/main")                              # type: ignore
+
+        # Disable and re-enable
+        cms_parser.disable()
+        cms_parser.enable()
+
+        # Confirm the cms_parser is enabled
+        if not self.is_cms_parser_enabled:
+            raise Exception("Something went wrong, cms_parser is still disabled.")
+
+        self.logger.info(
+            f"[{self.infrastructure.instance_name}] cms_parser successfully toggled to force run"
+        )
+
+    def wait_for_cms_main(self) -> None:
+        """
+        Checks the cms_main index until it has the expected number of events, or it times out.
```

Comment:
Why not raise an exception after timeout instead of just returning?

Reply:
Good question. I was trying to isolate validation into a single function; the ethos of this function is just to wait the acceptable amount of time. The ethos of `validate_against_cms` is to raise any validation errors.

Open to refactoring though

Reply:
Makes sense, resolving this one